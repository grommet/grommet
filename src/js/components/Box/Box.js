import React, {
  Children,
  forwardRef,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { ThemeContext } from 'styled-components';
import {
  backgroundIsDark,
  useForwardedRef,
  deviceResponsive,
  getBreakpoint,
} from '../../utils';
import { Keyboard } from '../Keyboard';

import { StyledBox, StyledBoxGap } from './StyledBox';
import { BoxPropTypes } from './propTypes';
import { SkeletonContext, useSkeleton } from '../Skeleton';
import { AnnounceContext } from '../../contexts/AnnounceContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { ResponsiveContainerContext, ResponsiveContext } from '../../contexts';
import { useThemeValue } from '../../utils/useThemeValue';
import { supportsContainerQueries } from '../../utils/responsive';

// The value of ResponsiveContainerContext
// when we're within a responsive container
const responsiveContainerValue = true;

const Box = forwardRef(
  (
    {
      a11yTitle,
      background: backgroundProp,
      border,
      children,
      cssGap,
      direction = 'column',
      elevation, // munged to avoid styled-components putting it in the DOM
      fill, // munged to avoid styled-components putting it in the DOM
      gap,
      kind, // munged to avoid styled-components putting it in the DOM
      onBlur,
      onClick,
      onFocus,
      overflow, // munged to avoid styled-components putting it in the DOM
      responsive: responsiveProp = true,
      tag,
      as,
      wrap, // munged to avoid styled-components putting it in the DOM,
      width, // munged to avoid styled-components putting it in the DOM
      height, // munged to avoid styled-components putting it in the DOM
      tabIndex,
      skeleton: skeletonProp,
      ...rest
    },
    ref,
  ) => {
    const { theme, passThemeFlag } = useThemeValue();
    // boxOptions was created to preserve backwards compatibility but
    // should not be supported in v3
    const { box: boxOptions } = useContext(OptionsContext);

    console.log('as', as, typeof as);
    const skeleton = useSkeleton();

    const responsiveContainer = useContext(ResponsiveContainerContext);
    const responsive =
      responsiveContainer && responsiveProp ? 'container' : responsiveProp;

    let background = backgroundProp;

    const announce = useContext(AnnounceContext);

    const [containerBreakpoint, setContainerBreakpoint] = useState(
      deviceResponsive(navigator.userAgent, theme) ||
        theme?.global?.deviceBreakpoints.tablet,
    );

    const containerRef = useForwardedRef(ref);
    // console.log('render', responsiveProp, containerRef);

    // Update the size in the ResponsiveContext at this
    // level if we're in responsive container mode.
    useEffect(() => {
      // console.log('usEffect', responsiveProp, containerRef);
      const containerElement = containerRef?.current;
      if (responsiveProp !== 'container' || !containerElement) return undefined;

      let resizeObserver;
      if (typeof window !== 'undefined' && window.ResizeObserver) {
        resizeObserver = new window.ResizeObserver((entries) => {
          window.requestAnimationFrame(() => {
            if (!Array.isArray(entries) || !entries.length) {
              return;
            }
            const size = entries[0]?.borderBoxSize?.[0]?.inlineSize;
            // console.log('Observer size', size, getBreakpoint(size, theme));
            if (size) {
              setContainerBreakpoint(getBreakpoint(size, theme));
            }
          });
        });

        resizeObserver.observe(containerElement);
      }
      // Initial size for ResponsiveContext in responsive container mode.
      // Also is a fallback for server side rendering
      const containerWidth = containerElement.getBoundingClientRect().width;
      setContainerBreakpoint(getBreakpoint(containerWidth, theme));

      return () => {
        if (resizeObserver) {
          resizeObserver.disconnect();
        }
      };
    }, [theme, responsiveProp, containerRef]);

    useEffect(() => {
      if (skeletonProp?.message?.start) announce(skeletonProp.message.start);
      else if (typeof skeletonProp?.message === 'string')
        announce(skeletonProp.message);
      return () =>
        skeletonProp?.message?.end && announce(skeletonProp.message.end);
    }, [announce, skeletonProp]);

    const focusable = useMemo(
      () => onClick && !(tabIndex < 0),
      [onClick, tabIndex],
    );

    const [focus, setFocus] = useState();

    const clickProps = useMemo(() => {
      if (focusable) {
        return {
          onClick,
          onFocus: (event) => {
            setFocus(true);
            if (onFocus) onFocus(event);
          },
          onBlur: (event) => {
            setFocus(false);
            if (onBlur) onBlur(event);
          },
        };
      }
      const result = {};
      if (onBlur) result.onBlur = onBlur;
      if (onClick) result.onClick = onClick;
      if (onFocus) result.onFocus = onFocus;
      return result;
    }, [focusable, onClick, onFocus, onBlur]);

    const adjustedTabIndex = useMemo(() => {
      if (tabIndex !== undefined) return tabIndex;
      if (focusable) return 0;
      return undefined;
    }, [focusable, tabIndex]);

    if (
      (border === 'between' ||
        (border && border.side === 'between') ||
        (Array.isArray(border) && border.find((b) => b.side === 'between'))) &&
      !gap
    ) {
      console.warn('Box must have a gap to use border between');
    }

    let contents = children;
    if (
      gap &&
      gap !== 'none' &&
      (!(boxOptions?.cssGap || cssGap || typeof gap === 'object') ||
        // need this approach to show border between
        border === 'between' ||
        border?.side === 'between' ||
        (Array.isArray(border) && border.find((b) => b.side === 'between')))
    ) {
      // if border is an array, we need to extract the border between object
      const styledBoxGapBorder = Array.isArray(border)
        ? border.find((b) => b.side === 'between')
        : border;
      const boxAs = !as && tag ? tag : as;
      contents = [];
      let firstIndex;
      Children.forEach(children, (child, index) => {
        if (child) {
          if (firstIndex === undefined) {
            firstIndex = index;
          } else {
            contents.push(
              <StyledBoxGap
                // eslint-disable-next-line react/no-array-index-key
                key={`gap-${index}`}
                as={boxAs === 'span' ? boxAs : 'div'}
                gap={gap}
                directionProp={direction}
                responsive={responsive}
                border={styledBoxGapBorder}
                {...passThemeFlag}
              />,
            );
          }
        }
        contents.push(child);
      });
    }

    const nextSkeleton = useMemo(() => {
      // Decide if we need to add a new SkeletonContext. We need one if:
      //   1. skeleton info was set in a property OR
      //   2. there already is a SkeletonContext but this box has a
      //      background or border. This means the box probably is more
      //      distinguishable from the area around it.
      // We keep track of a depth so we know how to alternate backgrounds.
      if (skeletonProp || ((background || border) && skeleton)) {
        const depth = skeleton ? skeleton.depth + 1 : 0;
        return {
          ...skeleton,
          depth,
          ...(typeof skeletonProp === 'object' ? skeletonProp : {}),
        };
      }
      return undefined;
    }, [background, border, skeleton, skeletonProp]);

    let skeletonProps = {};
    if (nextSkeleton) {
      const {
        colors: skeletonThemeColors,
        size: skeletonThemeSize,
        ...skeletonThemeProps
      } = theme.skeleton;
      const skeletonColors = nextSkeleton.colors
        ? nextSkeleton.colors[theme.dark ? 'dark' : 'light']
        : skeletonThemeColors?.[theme.dark ? 'dark' : 'light'];
      skeletonProps = { ...skeletonThemeProps };
      background = skeletonColors[nextSkeleton.depth % skeletonColors.length];
      if (skeletonProp?.animation) {
        skeletonProps.animation = skeletonProp.animation;
      }
      contents = (
        <SkeletonContext.Provider value={nextSkeleton}>
          {contents}
        </SkeletonContext.Provider>
      );
    }

    // construct a new theme object in case we have a background that wants
    // to change the background color context
    const nextTheme = useMemo(() => {
      let result;
      if (background || theme.darkChanged) {
        const dark = backgroundIsDark(background, theme);
        const darkChanged = dark !== undefined && dark !== theme.dark;
        if (darkChanged || theme.darkChanged) {
          result = { ...theme };
          result.dark = dark === undefined ? theme.dark : dark;
          result.background = background;
        } else if (background) {
          // This allows DataTable to intelligently set the background
          // of a pinned header or footer.
          result = { ...theme };
          result.background = background;
        }
      }
      return result || theme;
    }, [background, theme]);

    let content = (
      <StyledBox
        as={!as && tag ? tag : as}
        aria-label={a11yTitle}
        background={background}
        border={border}
        ref={containerRef}
        directionProp={direction}
        elevationProp={elevation}
        fillProp={fill}
        focus={focus}
        gap={
          (boxOptions?.cssGap || cssGap || typeof gap === 'object') &&
          gap &&
          gap !== 'none' &&
          border !== 'between' &&
          border?.side !== 'between' &&
          (!Array.isArray(border) ||
            !border.find((b) => b.side === 'between')) &&
          gap
        }
        kindProp={kind}
        overflowProp={overflow}
        wrapProp={wrap}
        widthProp={width}
        heightProp={height}
        responsive={responsive}
        responsiveContainer={responsiveProp === 'container'}
        tabIndex={adjustedTabIndex}
        {...clickProps}
        {...passThemeFlag}
        {...rest}
        {...skeletonProps}
      >
        <ThemeContext.Provider value={nextTheme}>
          {contents}
        </ThemeContext.Provider>
      </StyledBox>
    );

    if (responsiveProp === 'container') {
      if (supportsContainerQueries()) {
        content = (
          <ResponsiveContext.Provider value={containerBreakpoint}>
            <ResponsiveContainerContext.Provider
              value={responsiveContainerValue}
            >
              {content}
            </ResponsiveContainerContext.Provider>
          </ResponsiveContext.Provider>
        );
      } else {
        console.warn(
          '<Box responsive="container"> requires styled-components v6 or later',
        );
      }
    }

    if (onClick) {
      content = <Keyboard onEnter={onClick}>{content}</Keyboard>;
    }

    return content;
  },
);

Box.displayName = 'Box';
Box.propTypes = BoxPropTypes;
export { Box };
