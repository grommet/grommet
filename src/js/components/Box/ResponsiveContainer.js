import React, { useContext, useEffect, useRef, useState } from 'react';
import {
  ResponsiveContainerContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';
import { StyledResponsiveContainer } from './StyledResponsiveContainer';
import { deviceResponsive, getBreakpoint } from '../../utils';

const responsiveContainerValue = true;

export const ResponsiveContainer = ({ children }) => {
  const theme = useContext(ThemeContext);
  const ref = useRef(undefined);
  const [stateResponsive, setResponsive] = useState();

  useEffect(() => {
    const element = ref?.current;

    if (!element) return undefined;

    let resizeObserver;
    if (typeof window !== 'undefined' && window.ResizeObserver) {
      resizeObserver = new window.ResizeObserver((entries) => {
        window.requestAnimationFrame(() => {
          if (!Array.isArray(entries) || !entries.length) {
            return;
          }
          const size = entries[0]?.borderBoxSize?.[0]?.inlineSize;
          if (size) {
            setResponsive(getBreakpoint(size, theme));
          }
        });
      });

      resizeObserver.observe(element);
    } else {
      // fallback for server side rendering
      const { width } = element.getBoundingClientRect();
      setResponsive(getBreakpoint(width, theme));
    }
    return () => {
      if (resizeObserver) {
        resizeObserver.disconnect();
      }
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(navigator.userAgent, theme) ||
    theme?.global?.deviceBreakpoints.tablet;

  return (
    <ResponsiveContext.Provider value={responsive}>
      <ResponsiveContainerContext.Provider value={responsiveContainerValue}>
        <StyledResponsiveContainer ref={ref}>
          {children}
        </StyledResponsiveContainer>
      </ResponsiveContainerContext.Provider>
    </ResponsiveContext.Provider>
  );
};
