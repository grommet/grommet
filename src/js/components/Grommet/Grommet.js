import React, { forwardRef, useEffect, useMemo, useRef, useState } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ContainerTargetContext,
  ResponsiveContext,
  ThemeContext,
} from '../../contexts';

import {
  deepMerge,
  backgroundIsDark,
  deviceResponsive,
  getBreakpoint,
  normalizeColor,
  useForwardedRef,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';
import { RootsContext } from '../../contexts/RootsContext';
import { OptionsContext } from '../../contexts/OptionsContext';
import { format, MessageContext } from '../../contexts/MessageContext';
import defaultMessages from '../../languages/default.json';
import { GrommetPropTypes } from './propTypes';
import { AnalyticsProvider } from '../../contexts/AnalyticsContext';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

const defaultOptions = {};

const Grommet = forwardRef((props, ref) => {
  const {
    children,
    full,
    containerTarget = typeof document === 'object' ? document.body : undefined,
    theme: themeProp,
    options = defaultOptions,
    messages: messagesProp,
    onAnalytics,
    ...rest
  } = props;
  const { background, dir, themeMode, userAgent } = props;
  const [stateResponsive, setResponsive] = useState();

  const theme = useMemo(() => {
    const nextTheme = deepMerge(baseTheme, themeProp || {});

    // if user provides specific menu alignment, we don't want
    // the defaults to be included at all (can cause issues with controlMirror)
    // override merged value with themeProp value
    if (
      themeProp &&
      themeProp.menu &&
      themeProp.menu.drop &&
      themeProp.menu.drop.align
    ) {
      delete nextTheme.menu.drop.align;
      nextTheme.menu.drop.align = themeProp.menu.drop.align;
    }
    const {
      colors: { background: themeBackground },
    } = nextTheme.global;

    nextTheme.dark = (themeMode || nextTheme.defaultMode) === 'dark';

    if (
      themeMode === 'auto' &&
      typeof window !== 'undefined' &&
      window.matchMedia &&
      window.matchMedia('(prefers-color-scheme: dark)').matches
    ) {
      nextTheme.dark = true;
    }

    const color = normalizeColor(background || themeBackground, nextTheme);
    nextTheme.dark = backgroundIsDark(color, nextTheme);
    nextTheme.baseBackground = background || themeBackground;
    // This allows DataTable to intelligently set the background of a pinned
    // header or footer.
    nextTheme.background = nextTheme.baseBackground;

    if (dir) {
      nextTheme.dir = dir;
    }

    return nextTheme;
  }, [background, dir, themeMode, themeProp]);

  const messages = useMemo(() => {
    // combine the passed in messages, if any, with the default
    // messages and format function.
    const nextMessages = deepMerge(
      defaultMessages,
      messagesProp?.messages || {},
    );
    return {
      messages: nextMessages,
      format: (opts) => {
        const message = messagesProp?.format && messagesProp.format(opts);
        return typeof message !== 'undefined'
          ? message
          : format(opts, nextMessages);
      },
    };
  }, [messagesProp]);

  useEffect(() => {
    const onResize = () => {
      setResponsive(getBreakpoint(document.body.clientWidth, theme));
    };
    window.addEventListener('resize', onResize);
    onResize();
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [theme]);

  const responsive =
    stateResponsive ||
    deviceResponsive(userAgent, theme) ||
    theme.global.deviceBreakpoints.tablet;

  const grommetRef = useForwardedRef(ref);

  // track open FocusedContainers in a global array to manage
  // focus event listeners for trapFocus
  const roots = useRef([]);
  useEffect(() => {
    if (grommetRef.current) roots.current.push(grommetRef.current);
  }, [grommetRef]);
  const rootsContextValue = useMemo(() => ({ roots }), []);

  return (
    <ThemeContext.Provider value={theme}>
      <ResponsiveContext.Provider value={responsive}>
        <RootsContext.Provider value={rootsContextValue}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <OptionsContext.Provider value={options}>
              <MessageContext.Provider value={messages}>
                <AnalyticsProvider onAnalytics={onAnalytics}>
                  <StyledGrommet full={full} {...rest} ref={grommetRef}>
                    {children}
                  </StyledGrommet>
                  {full && <FullGlobalStyle />}
                </AnalyticsProvider>
              </MessageContext.Provider>
            </OptionsContext.Provider>
          </ContainerTargetContext.Provider>
        </RootsContext.Provider>
      </ResponsiveContext.Provider>
    </ThemeContext.Provider>
  );
});

Grommet.displayName = 'Grommet';
Grommet.propTypes = GrommetPropTypes;

export { Grommet };
