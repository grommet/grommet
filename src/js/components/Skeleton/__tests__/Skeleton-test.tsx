import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import { hpe } from 'grommet-theme-hpe';

import { Skeleton } from '..';
import { Grommet } from '../../Grommet';
import { Box } from '../../Box';
import { Text } from '../../Text';
import { Heading } from '../../Heading';
import { Paragraph } from '../../Paragraph';
import { Button } from '../../Button';

describe('Skeleton', () => {
  test('renders', () => {
    const { asFragment } = render(
      <Grommet theme={hpe}>
        <Skeleton />
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('renders outside grommet wrapper', () => {
    const { asFragment } = render(<Skeleton />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Box skeleton loading', () => {
    const { asFragment } = render(
      <Grommet theme={hpe}>
        <Box skeleton>
          <Heading>Heading</Heading>
          <Text>Text</Text>
          <Paragraph>Paragraph</Paragraph>
          <Button label="Button" onClick={() => {}} />
        </Box>
      </Grommet>,
    );
    expect(screen.queryByText('Heading')).toBeNull();
    expect(screen.queryByText('Text')).toBeNull();
    expect(screen.queryByText('Paragraph')).toBeNull();
    expect(screen.queryByText('Button')).toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  test('Box skeleton sizes loading', () => {
    const { asFragment } = render(
      <Grommet theme={hpe}>
        <Box skeleton>
          <Heading level={2} size="small">
            Heading
          </Heading>
          <Text size="xsmall">Text</Text>
          <Paragraph size="large">Paragraph</Paragraph>
          <Button size="large" label="Button" onClick={() => {}} />
        </Box>
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByText('Heading')).toBeNull();
  });

  test('Box skeleton loaded', () => {
    const { asFragment } = render(
      <Grommet theme={hpe}>
        <Box skeleton={false}>
          <Heading>Heading</Heading>
          <Text>Text</Text>
          <Paragraph>Paragraph</Paragraph>
          <Button label="Button" onClick={() => {}} />
        </Box>
      </Grommet>,
    );
    expect(screen.queryByText('Heading')).not.toBeNull();
    expect(screen.queryByText('Text')).not.toBeNull();
    expect(screen.queryByText('Paragraph')).not.toBeNull();
    expect(screen.queryByText('Button')).not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  test('Box skeleton with specific dimensions', () => {
    const { asFragment } = render(
      <Grommet theme={hpe}>
        <Box skeleton height="small" width="medium">
          <Text>text</Text>
        </Box>
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('with styling props', () => {
    const { asFragment } = render(
      <Grommet>
        <Skeleton
          height="small"
          margin="small"
          round="small"
          colors={{ dark: ['green', 'blue'], light: ['pink', 'orange'] }}
        />
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('Skeleton with theme', () => {
    const { asFragment } = render(
      <Grommet
        theme={{
          skeleton: {
            colors: {
              light: ['#a2a8a8', '#adb9ba'],
            },
            round: 'xsmall',
            extend: 'border: 1px solid blue;',
          },
        }}
      >
        <Box skeleton>
          <Text>Text</Text>
        </Box>
      </Grommet>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
