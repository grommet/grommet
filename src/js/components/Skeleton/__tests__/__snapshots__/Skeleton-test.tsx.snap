// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Skeleton Box skeleton loaded 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Metric',Arial,sans-serif;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: #ffffff;
  color: #555555;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c3 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  font-weight: undefined;
}

.c2 {
  font-size: 2.25rem;
  line-height: 2.5rem;
  max-width: 1200px;
  font-weight: 400;
  overflow-wrap: break-word;
  color: #2e2e2e;
}

.c4 {
  font-size: 1.125rem;
  line-height: 1.5rem;
  max-width: 30em;
  font-weight: undefined;
}

.c5 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: none;
  border-radius: 9999px;
  padding: 5px 18px;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: rgba(0, 0, 0, 0);
  color: #2e2e2e;
  border-style: solid;
  border-width: 1px;
  border-color: rgba(0, 0, 0, 0);
  border-radius: 9999px;
  font-weight: 500;
  padding: 5px 18px;
  text-align: center;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c5:hover {
  background-color: rgba(0, 0, 0, 0.04);
  color: #2e2e2e;
  border-color: rgba(0, 0, 0, 0);
  font-weight: 500;
}

.c5:focus {
  outline-offset: 2px;
  outline: 2px solid #004233;
  box-shadow: 0 0 2px 2px #ffffff;
}

.c5:focus >circle,
.c5:focus >ellipse,
.c5:focus >line,
.c5:focus >path,
.c5:focus >polygon,
.c5:focus >polyline,
.c5:focus >rect {
  outline-offset: 2px;
  outline: 2px solid #004233;
  box-shadow: 0 0 2px 2px #ffffff;
}

.c5:focus ::-moz-focus-inner {
  border: 0;
}

.c5:focus:not(:focus-visible) {
  outline-offset: 0px;
  outline: none;
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible) >circle,
.c5:focus:not(:focus-visible) >ellipse,
.c5:focus:not(:focus-visible) >line,
.c5:focus:not(:focus-visible) >path,
.c5:focus:not(:focus-visible) >polygon,
.c5:focus:not(:focus-visible) >polyline,
.c5:focus:not(:focus-visible) >rect {
  outline-offset: 0px;
  outline: none;
  outline: none;
  box-shadow: none;
}

.c5:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

@media only screen and (max-width: 768px) {
  .c2 {
    font-size: 1.75rem;
    line-height: 2rem;
    max-width: 816px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h1
        class="c2"
      >
        Heading
      </h1>
      <span
        class="c3"
      >
        Text
      </span>
      <p
        class="c4"
      >
        Paragraph
      </p>
      <button
        class="c5"
        type="button"
      >
        Button
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Skeleton Box skeleton loading 1`] = `
<DocumentFragment>
  .c3 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 2.5rem;
  max-width: 200px;
  min-width: 150px;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  margin-top: 6px;
  margin-bottom: 6px;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.125rem;
  min-width: 100px;
  width: 100px;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.125rem;
  width: 100%;
}

.c8 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.125rem;
  width: 30%;
}

.c9 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.5rem;
  min-width: 100px;
  padding-left: 18px;
  padding-right: 18px;
  padding-top: 5px;
  padding-bottom: 5px;
}

.c0 {
  font-family: 'Metric',Arial,sans-serif;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: #ffffff;
  color: #555555;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #ffffff;
  color: #555555;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  max-width: 30em;
}

.c7 {
  flex: 0 0 auto;
  align-self: stretch;
  height: 6px;
}

.c2 {
  font-size: 2.25rem;
  line-height: 2.5rem;
  max-width: 1200px;
  font-weight: 400;
  overflow-wrap: break-word;
  color: #2e2e2e;
}

@media only screen and (max-width: 768px) {
  .c3 {
    height: 2rem;
  }
}

@media only screen and (max-width: 768px) {
  .c7 {
    height: 3px;
  }
}

@media only screen and (max-width: 768px) {
  .c2 {
    font-size: 1.75rem;
    line-height: 2rem;
    max-width: 816px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h1
        class="c2"
      >
        <div
          aria-hidden="true"
          class="c3"
        />
      </h1>
      <div
        aria-hidden="true"
        class="c4"
      />
      <div
        class="c5"
      >
        <div
          aria-hidden="true"
          class="c6"
        />
        <div
          class="c7"
        />
        <div
          aria-hidden="true"
          class="c6"
        />
        <div
          class="c7"
        />
        <div
          aria-hidden="true"
          class="c8"
        />
      </div>
      <div
        aria-hidden="true"
        class="c9"
        default=""
        option="[object Object]"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Skeleton Box skeleton sizes loading 1`] = `
<DocumentFragment>
  .c3 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.75rem;
  max-width: 200px;
  min-width: 150px;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  margin-top: 6px;
  margin-bottom: 6px;
  background-color: #f7f7f7;
  color: #555555;
  height: 0.875rem;
  min-width: 100px;
  width: 100px;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.375rem;
  width: 100%;
}

.c8 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.375rem;
  width: 30%;
}

.c9 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.75rem;
  min-width: 100px;
  padding-left: 24px;
  padding-right: 24px;
  padding-top: 9px;
  padding-bottom: 9px;
}

.c0 {
  font-family: 'Metric',Arial,sans-serif;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: #ffffff;
  color: #555555;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #ffffff;
  color: #555555;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c5 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  max-width: 30em;
}

.c7 {
  flex: 0 0 auto;
  align-self: stretch;
  height: 6px;
}

.c2 {
  font-size: 1.5rem;
  line-height: 1.75rem;
  max-width: 624px;
  font-weight: 400;
  overflow-wrap: break-word;
  color: #2e2e2e;
}

@media only screen and (max-width: 768px) {
  .c3 {
    height: 1.5rem;
  }
}

@media only screen and (max-width: 768px) {
  .c7 {
    height: 3px;
  }
}

@media only screen and (max-width: 768px) {
  .c2 {
    font-size: 1.25rem;
    line-height: 1.5rem;
    max-width: 528px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h2
        class="c2"
      >
        <div
          aria-hidden="true"
          class="c3"
        />
      </h2>
      <div
        aria-hidden="true"
        class="c4"
      />
      <div
        class="c5"
      >
        <div
          aria-hidden="true"
          class="c6"
        />
        <div
          class="c7"
        />
        <div
          aria-hidden="true"
          class="c6"
        />
        <div
          class="c7"
        />
        <div
          aria-hidden="true"
          class="c8"
        />
      </div>
      <div
        aria-hidden="true"
        class="c9"
        default=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Skeleton Box skeleton with specific dimensions 1`] = `
<DocumentFragment>
  .c2 {
  display: flex;
  box-sizing: border-box;
  margin-top: 6px;
  margin-bottom: 6px;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.125rem;
  min-width: 100px;
  width: 100px;
}

.c0 {
  font-family: 'Metric',Arial,sans-serif;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: #ffffff;
  color: #555555;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #ffffff;
  color: #555555;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 192px;
  width: 384px;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        aria-hidden="true"
        class="c2"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Skeleton Skeleton with theme 1`] = `
<DocumentFragment>
  .c2 {
  display: flex;
  box-sizing: border-box;
  margin-top: 6px;
  margin-bottom: 6px;
  background-color: #EDEDED;
  color: #444444;
  height: 18px;
  min-width: 100px;
  width: 100px;
  border: 1px solid blue;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  background-color: #a2a8a8;
  color: #444444;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  border-radius: 6px;
}

@media only screen and (max-width: 768px) {
  .c1 {
    border-radius: 3px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        aria-hidden="true"
        class="c2"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Skeleton renders 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  box-sizing: border-box;
  background-color: #f7f7f7;
  color: #555555;
  height: 1.5rem;
  width: 100%;
}

.c0 {
  font-family: 'Metric',Arial,sans-serif;
  font-size: 1.125rem;
  line-height: 1.5rem;
  background-color: #ffffff;
  color: #555555;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
    class="c0"
  >
    <div
      aria-hidden="true"
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`Skeleton renders outside grommet wrapper 1`] = `
<DocumentFragment>
  .c0 {
  display: flex;
  box-sizing: border-box;
  background-color: #EDEDED;
  color: #444444;
  height: 24px;
  width: 100%;
}

<div
    aria-hidden="true"
    class="c0"
  />
</DocumentFragment>
`;

exports[`Skeleton with styling props 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  box-sizing: border-box;
  margin: 12px;
  background: orange;
  height: 192px;
  width: 100%;
  border-radius: 12px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
    class="c0"
  >
    <div
      aria-hidden="true"
      class="c1"
    />
  </div>
</DocumentFragment>
`;
